File: tools/stress/stress.py
#!/usr/bin/env python3
"""Resource-limit stress for PageBook on testnet (market 0).

Each account holds 40 resting quotes dispersed across all 32 tick words of
market 0's band, and re-quotes all 40 in one `replace_batch` as fast as the
chain admits it. With the section 14 pad, one such batch declares close to
the 200 read-write-entry per-transaction cap and 40 to 60 KB of write bytes:
the largest legal writer the contract has. N accounts firing together
oversubscribe the per-ledger caps (1,000 write entries, 286,720 write bytes),
which is the point: watch which cap rations inclusion and how.

Phases:
  seed    rest the 40 quotes per account (idempotent: skips live nonces)
  run     fire batches for --ledgers, log every outcome with submit ledger
  analyze read the log, join with getTransaction ledgers, print per-ledger
          admitted totals and inclusion delay percentiles

  python3 tools/stress/stress.py seed|run|analyze --accounts 8 [--ledgers 250]
"""
import argparse
import collections
import concurrent.futures
import json
import os
import random
import sys
import threading
import time

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "soak"))
import soak  # noqa: E402

CONTRACT = "CDX3WVFY6GV53J3XT53MNPE5HVKAGTCH74W3AWGMI43KUFK5TSXOU2RO"
MARKET = 0
BASE_SAC = "CDAHSKHBGFENTV3XGWRWVIWE3ISAEYIZQNGD4GCWRDDIOIW4DVZ26FQG"
QUOTE_SAC = "CBEC6J5RWWWC7CYCHJTXIBDFTFRK6GTMLK4E47BECO5BDXVM7YHATUIK"
ISSUER = "GCBMNFRU74KLBUCVHJVQXRRMGEWUWC2WZ5KXLYABNFLXGCFTJPKBT4IB"
CODES = ["PBA", "PBB"]
NONCE_BASE = 777_000_000


def ticks_for(acct_i):
    """40 ask ticks for account i: one per word 0..31, plus 8 more in words
    0..7; offset per account so accounts do not share levels."""
    ts = [2048 * w + 200 + acct_i * 17 for w in range(32)]
    ts += [2048 * w + 1200 + acct_i * 17 for w in range(8)]
    return ts


class Stress:
    def __init__(self, a):
        self.a = a
        self.cli = soak.Cli(CONTRACT, a.config_dir, a.network, a.rpc)
        self.log = open(a.log, "a")
        self.lock = threading.Lock()
        self.counts = collections.Counter()

    def record(self, **kw):
        kw["t"] = time.time()
        with self.lock:
            self.counts[kw.get("outcome", "?")] += 1
            self.log.write(json.dumps(kw) + "\n")
            self.log.flush()

    def token_keys(self, addr):
        keys = []
        for sac, code in ((BASE_SAC, "PBA"), (QUOTE_SAC, "PBB")):
            keys.append({"contract_data": {"contract": sac, "key": "ledger_key_contract_instance", "durability": "persistent"}})
            keys.append({"contract_data": {"contract": sac, "key": {"vec": [{"symbol": "Balance"}, {"address": CONTRACT}]}, "durability": "persistent"}})
            keys.append({"trustline": {"account_id": addr, "asset": {"credit_alphanum4": {"asset_code": code, "issuer": ISSUER}}}})
        keys.append(soak.ck(CONTRACT, "FeeAccrual", MARKET, BASE_SAC))
        keys.append(soak.ck(CONTRACT, "FeeAccrual", MARKET, QUOTE_SAC))
        return keys

    def submit(self, source, fn, args, pad, **extra):
        addr = extra.pop("addr", None)
        t0 = time.time()
        try:
            l0 = self.cli.latest_ledger()
        except Exception:
            l0 = None
        try:
            xdr = self.cli.build(source, fn, args)
            sim = self.cli.simulate(source, xdr)
        except soak.SimError as e:
            e = str(e)
            self.record(acct=source, action=fn, outcome="sim:" + soak.classify(1, "", e), detail=e[-260:], **extra)
            return None
        except Exception as e:
            self.record(acct=source, action=fn, outcome="build_error", detail=str(e)[-260:], **extra)
            return None
        obj = self.cli.decode(sim)
        added = soak.apply_pad(obj, pad)
        sd = obj["tx"]["tx"]["ext"]["v1"]
        res = sd["resources"]
        decl = {"rw": len(res["footprint"]["read_write"]), "ro": len(res["footprint"]["read_only"]),
                "instr": int(res["instructions"]), "wb": int(res["write_bytes"]), "fee": int(sd["resource_fee"])}
        final = self.cli.encode(obj)
        try:
            signed = self.cli.sign(source, final)
        except Exception as e:
            self.record(acct=source, action=fn, outcome="sign_error", detail=str(e)[-260:], **extra)
            return None
        rc, out, err = self.cli.send(signed)
        outcome = soak.classify(rc, out, err)
        text = out + err
        import re
        m = re.search(r"Transaction hash is ([0-9a-f]{64})", text)
        h = m.group(1) if m else ""
        if outcome in ("other", "trapped") and "Trapped" in text and h:
            outcome = self.cli.diagnose(h)
        self.record(acct=source, action=fn, outcome=outcome, tx=h, submit_ledger=l0, wall_s=round(time.time() - t0, 2),
                    detail=text[-200:] if outcome != "ok" else "", **decl, **extra)
        return outcome

    # ---- phases
    def seed(self):
        with concurrent.futures.ThreadPoolExecutor(max_workers=8) as ex:
            list(ex.map(self._seed_one, range(self.a.accounts)))

    def _seed_one(self, i):
        if True:
            src = f"pb-stress{i+1}"
            addr = self.cli.address(src)
            for j, tick in enumerate(ticks_for(i)):
                nonce = NONCE_BASE + i * 1000 + j
                try:
                    self.cli.invoke_readonly(src, "order", ["--market", str(MARKET), "--owner", addr, "--nonce", str(nonce)])
                    continue  # already live
                except Exception:
                    pass
                window = json.dumps({"consume": [], "append": {"first": 0, "last": 1}})
                flags = json.dumps({"post_only": True, "fill_or_kill": False, "no_rest": False})
                args = ["--taker", addr, "--market", str(MARKET), "--is_bid", "false", "--limit_tick", str(tick),
                        "--qty_lots", "2", "--start_tick", str(65535), "--nonce", str(nonce), "--window", window, "--flags", flags]
                pad = soak.rest_keys(CONTRACT, MARKET, False, tick) + [soak.order_key(CONTRACT, MARKET, addr, nonce)] + self.token_keys(addr)
                self.submit(src, "place", args, pad, addr=addr, seed=1)
            print(f"{src} seeded", file=sys.stderr)

    def batch_once(self, i):
        src = f"pb-stress{i+1}"
        addr = self.cli.address(src)
        items = []
        pad = self.token_keys(addr)
        for j, tick in enumerate(ticks_for(i)):
            nonce = NONCE_BASE + i * 1000 + j
            items.append({"nonce": nonce, "is_bid": False, "tick": tick, "qty_lots": random.randint(2, 5),
                          "window": {"consume": [], "append": {"first": 0, "last": 1}}})
            # own-side keys per item; page 0 only (queues stay inline at 2-5 lots)
            pad.append(soak.ck(CONTRACT, "Level", MARKET, False, tick))
            pad.append(soak.ck(CONTRACT, "TickWord", MARKET, False, soak.word_of(tick)))
            pad.append(soak.ck(CONTRACT, "LevelPage", MARKET, False, tick, 0))
        pad.append(soak.ck(CONTRACT, "TickSummary", MARKET, False))
        pad.append(soak.ck(CONTRACT, "BestTick", MARKET, False))
        pad.append(soak.ck(CONTRACT, "BestTick", MARKET, True))
        # top up with a neighbor account's existing levels to sit just under the
        # 200 read-write cap (existing entries also cost declared write bytes)
        for tick in ticks_for((i + 1) % self.a.accounts)[:self.a.extra_pad]:
            pad.append(soak.ck(CONTRACT, "Level", MARKET, False, tick))
        return self.submit(src, "replace_batch", ["--owner", addr, "--market", str(MARKET), "--items", json.dumps(items)], pad, addr=addr)

    def run(self):
        start = self.cli.latest_ledger()
        end = start + self.a.ledgers
        stop = threading.Event()

        def worker(i):
            while not stop.is_set():
                self.batch_once(i)
                time.sleep(self.a.pause)

        threads = [threading.Thread(target=worker, args=(i,), daemon=True) for i in range(self.a.accounts)]
        for t in threads:
            t.start()
        last = start
        while True:
            time.sleep(10)
            try:
                cur = self.cli.latest_ledger()
            except Exception:
                continue
            if cur >= end:
                break
            if cur - last >= 30:
                last = cur
                self.record(acct="stress", action="progress", outcome="tick", detail=f"ledger {cur}/{end} counts={json.dumps(dict(self.counts))}")
        stop.set()
        time.sleep(20)
        self.record(acct="stress", action="done", outcome="summary", detail=json.dumps({"start": start, "end": end, "counts": dict(self.counts)}))
        print(json.dumps({"start": start, "end": end, "counts": dict(self.counts)}, indent=1))

    def analyze(self):
        recs = []
        for line in open(self.a.log):
            try:
                d = json.loads(line)
            except Exception:
                continue
            if d.get("action") == "replace_batch" and d.get("outcome") == "ok" and d.get("tx"):
                recs.append(d)
        print(f"{len(recs)} landed batches", file=sys.stderr)

        def ledger_of(r):
            try:
                g = self.cli.rpc_call("getTransaction", {"hash": r["tx"]})
                return g.get("ledger"), int(json.loads(self.cli._run(["xdr", "decode", "--type", "TransactionResult", "--output", "json"], stdin=g["resultXdr"])[1])["fee_charged"])
            except Exception:
                return None, None
        with concurrent.futures.ThreadPoolExecutor(max_workers=8) as ex:
            got = list(ex.map(ledger_of, recs))
        per = collections.defaultdict(lambda: {"n": 0, "rw": 0, "wb": 0, "instr": 0, "fee": 0})
        delays = []
        for r, (led, fee) in zip(recs, got):
            if led is None:
                continue
            p = per[led]
            p["n"] += 1
            p["rw"] += r["rw"]
            p["wb"] += r["wb"]
            p["instr"] += r["instr"]
            p["fee"] += fee or 0
            if r.get("submit_ledger"):
                delays.append(led - r["submit_ledger"])
        out = {"ledgers_with_batches": len(per),
               "per_ledger": {"n": self._q([p["n"] for p in per.values()]),
                              "rw_declared": self._q([p["rw"] for p in per.values()]),
                              "write_bytes_declared": self._q([p["wb"] for p in per.values()]),
                              "instr_declared": self._q([p["instr"] for p in per.values()])},
               "inclusion_delay_ledgers": self._q(delays),
               "wall_s": self._q([r["wall_s"] for r in recs]),
               "fee_charged": self._q([f for (_, f) in got if f])}
        print(json.dumps(out, indent=1))
        top = sorted(per.items(), key=lambda kv: -kv[1]["wb"])[:10]
        print("busiest ledgers (by our declared write bytes):")
        for led, p in top:
            print(f"  {led}: {p['n']} batches, rw {p['rw']} ({p['rw']/10:.0f}% of cap), wb {p['wb']:,} ({p['wb']/2867.2:.0f}% of cap), instr {p['instr']/1e6:.0f}M ({p['instr']/5.8e6:.0f}% of cap)")

    @staticmethod
    def _q(v):
        if not v:
            return None
        v = sorted(v)
        def q(p):
            i = (len(v) - 1) * p
            lo = int(i)
            return v[lo] if lo == len(v) - 1 else v[lo] + (v[lo + 1] - v[lo]) * (i - lo)
        return {"min": v[0], "p50": round(q(.5), 2), "p95": round(q(.95), 2), "max": v[-1]}


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("phase", choices=["seed", "run", "analyze"])
    ap.add_argument("--accounts", type=int, default=8)
    ap.add_argument("--ledgers", type=int, default=250)
    ap.add_argument("--pause", type=float, default=1.0)
    ap.add_argument("--extra-pad", type=int, default=28)
    ap.add_argument("--config-dir", default=".stellar")
    ap.add_argument("--network", default="testnet")
    ap.add_argument("--rpc", default="https://soroban-testnet.stellar.org")
    ap.add_argument("--log", default="tools/stress/stress.log")
    a = ap.parse_args()
    s = Stress(a)
    getattr(s, a.phase)()


if __name__ == "__main__":
    main()

File: clients/web/ops/stress.ts
import { resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { createRpc, type Rpc } from "../src/book";
import { addrToHex, type ClientKey } from "../src/engine/clientKeys";
import { wordOf } from "../src/decode";
import { submitPostOnlyPlace, submitReplaceBatch, type ClassicToken } from "../src/engine/submit";
import type { WindowSpec } from "../src/engine/pad";
import { parseArgs, type ArgSpec } from "./lib/args";
import { loadIdentity, type Identity } from "./lib/identity";
import { latestLedger, waitLedgers } from "./lib/ledger";
import {
  MARKET0_BASE_SAC,
  MARKET0_CODES,
  MARKET0_CONTRACT,
  MARKET0_ID,
  MARKET0_ISSUER,
  MARKET0_NONCE_BASE,
  MARKET0_QUOTE_SAC,
} from "./lib/market0";
import { emptyRestWindow, randInt, repr } from "./lib/math";
import { openLog, type OpsLog } from "./lib/opslog";
import { parseLogLines } from "./lib/logparse";
import { outcomeOf, type OutcomeInput } from "./lib/outcomes";
import { classicPairTokens, feeKeys, orderClientKey, restKeys, tokenHex } from "./lib/padkeys";
import { mapPool } from "./lib/pool";
import { percentiles } from "./lib/stats";
import { resultText, sleep } from "./lib/submitlog";
import { decodeFeeCharged } from "./lib/txdecode";
import { createViews, type Views } from "./lib/views";
import { readFileSync } from "node:fs";

export const CONTRACT = MARKET0_CONTRACT;
export const MARKET = MARKET0_ID;
export const BASE_SAC = MARKET0_BASE_SAC;
export const QUOTE_SAC = MARKET0_QUOTE_SAC;
export const ISSUER = MARKET0_ISSUER;
export const CODES = MARKET0_CODES;
export const NONCE_BASE = MARKET0_NONCE_BASE;

export function ticksFor(acctI: number): number[] {
  const ts = Array.from({ length: 32 }, (_, w) => 2048 * w + 200 + acctI * 17);
  ts.push(...Array.from({ length: 8 }, (_, w) => 2048 * w + 1200 + acctI * 17));
  return ts;
}

export type StressPhase = "seed" | "run" | "analyze";

export type StressArgs = {
  phase: StressPhase;
  accounts: number;
  ledgers: number;
  pause: number;
  extraPad: number;
  configDir: string;
  network: string;
  rpc: string;
  log: string;
};

const STRESS_SPECS: ArgSpec<string>[] = [
  { flag: "--accounts", dest: "accounts", type: "int", default: 8 },
  { flag: "--ledgers", dest: "ledgers", type: "int", default: 250 },
  { flag: "--pause", dest: "pause", type: "float", default: 1.0 },
  { flag: "--extra-pad", dest: "extraPad", type: "int", default: 28 },
  { flag: "--config-dir", dest: "configDir", default: ".stellar" },
  { flag: "--network", dest: "network", default: "testnet" },
  { flag: "--rpc", dest: "rpc", default: "https://soroban-testnet.stellar.org" },
  { flag: "--log", dest: "log", default: "ops/stress.log" },
];

const PHASES = new Set<string>(["seed", "run", "analyze"]);

export function parseStressArgs(argv: string[]): StressArgs {
  const rest = [...argv];
  let phase: string | undefined;
  for (let i = 0; i < rest.length; i++) {
    if (!rest[i].startsWith("-") && PHASES.has(rest[i])) {
      phase = rest[i];
      rest.splice(i, 1);
      break;
    }
  }
  if (!phase) throw new Error("missing required argument: phase (seed|run|analyze)");
  const flags = parseArgs<Omit<StressArgs, "phase">>(rest, STRESS_SPECS as ArgSpec<keyof Omit<StressArgs, "phase"> & string>[]);
  return { ...flags, phase: phase as StressPhase };
}

export type StressDeps = {
  rpc?: Rpc;
  ids?: Record<string, Identity>;
  views?: Record<string, Views>;
  log?: OpsLog;
  now?: () => number;
  sleep?: (ms: number) => Promise<void>;
  rnd?: () => number;
  latest?: () => Promise<number>;
  readLog?: () => string;
};

export class Stress {
  a: StressArgs;
  rpc: Rpc;
  log: OpsLog;
  counts: Record<string, number> = {};
  tokens: ClassicToken[];
  hex: { base: ReturnType<typeof addrToHex>; quote: ReturnType<typeof addrToHex> };
  stop = false;
  private now: () => number;
  private sleepFn: (ms: number) => Promise<void>;
  private rnd: () => number;
  private latestFn: () => Promise<number>;
  private ids: Record<string, Identity>;
  private viewsCache: Record<string, Views>;
  private givenViews?: Record<string, Views>;
  private readLog: () => string;

  constructor(a: StressArgs, deps: StressDeps = {}) {
    this.a = a;
    this.rpc = deps.rpc ?? createRpc(a.rpc);
    this.log = deps.log ?? openLog(a.log);
    this.now = deps.now ?? (() => Date.now() / 1000);
    this.sleepFn = deps.sleep ?? sleep;
    this.rnd = deps.rnd ?? Math.random;
    this.latestFn = deps.latest ?? (() => latestLedger(this.rpc));
    this.ids = deps.ids ?? {};
    this.viewsCache = {};
    this.givenViews = deps.views;
    this.tokens = classicPairTokens(BASE_SAC, QUOTE_SAC, ISSUER, CODES.join(","));
    this.hex = tokenHex(BASE_SAC, QUOTE_SAC);
    this.readLog = deps.readLog ?? (() => readFileSync(a.log, "utf8"));
  }

  src(i: number): string {
    return `pb-stress${i + 1}`;
  }

  idOf(name: string): Identity {
    if (!this.ids[name]) this.ids[name] = loadIdentity(name, this.a.configDir);
    return this.ids[name];
  }

  viewsOf(name: string): Views {
    if (this.givenViews?.[name]) return this.givenViews[name];
    if (!this.viewsCache[name]) {
      const id = this.idOf(name);
      this.viewsCache[name] = createViews(this.rpc, {
        contract: CONTRACT,
        source: id.address,
        market: MARKET,
        owner: id.address,
      });
    }
    return this.viewsCache[name];
  }

  record(kw: Record<string, unknown>): void {
    const outcome = String(kw.outcome ?? "?");
    this.counts[outcome] = (this.counts[outcome] ?? 0) + 1;
    this.log.record(String(kw.action ?? ""), outcome, { ...kw, t: this.now() });
  }

  tokenPadKeys(): ClientKey[] {
    return feeKeys(MARKET, this.hex.base, this.hex.quote);
  }

  async submit(source: string, fn: string, run: () => Promise<OutcomeInput>, extra: Record<string, unknown> = {}): Promise<string | null> {
    const t0 = this.now();
    let l0: number | null = null;
    try {
      l0 = await this.latestFn();
    } catch {
      l0 = null;
    }
    let res: OutcomeInput;
    try {
      res = await run();
    } catch (e) {
      const msg = repr(e);
      const sim = /simulat/i.test(msg);
      this.record({
        acct: source,
        action: fn,
        outcome: sim ? outcomeOf({ kind: "rpc", message: msg, at: "simulation" }) : "build_error",
        detail: msg.slice(-260),
        ...extra,
      });
      return null;
    }
    const out = outcomeOf(res);
    const tx = "hash" in res ? (res as { hash?: string }).hash ?? "" : "";
    const decl = "declared" in res && res.declared ? res.declared : {};
    const text = resultText(res);
    this.record({
      acct: source,
      action: fn,
      outcome: out,
      tx,
      submit_ledger: l0,
      wall_s: Math.round((this.now() - t0) * 100) / 100,
      detail: out === "ok" ? "" : text.slice(-200),
      ...decl,
      ...extra,
    });
    return out;
  }

  async seed(): Promise<void> {
    await mapPool(
      Array.from({ length: this.a.accounts }, (_, i) => i),
      8,
      (i) => this.seedOne(i),
    );
  }

  async seedOne(i: number): Promise<void> {
    const src = this.src(i);
    const id = this.idOf(src);
    const views = this.viewsOf(src);
    const ownerHex = addrToHex(id.address);
    for (let j = 0; j < ticksFor(i).length; j++) {
      const tick = ticksFor(i)[j];
      const nonce = NONCE_BASE + i * 1000 + j;
      try {
        const live = await views.order(nonce);
        if (live) continue;
      } catch {
        /* place */
      }
      const padKeys: ClientKey[] = [
        ...restKeys(MARKET, false, tick),
        orderClientKey(MARKET, ownerHex, BigInt(nonce)),
        ...this.tokenPadKeys(),
      ];
      await this.submit(
        src,
        "place",
        () =>
          submitPostOnlyPlace(this.rpc, {
            contract: CONTRACT,
            secret: id.secret,
            taker: id.address,
            market: MARKET,
            isBid: false,
            limitTick: tick,
            qtyLots: 2n,
            startTick: 65535,
            nonce: BigInt(nonce),
            window: emptyRestWindow(),
            flags: { post_only: true, fill_or_kill: false, no_rest: false },
            padKeys,
            tokens: this.tokens,
          }),
        { seed: 1 },
      );
    }
    process.stderr.write(`${src} seeded\n`);
  }

  async batchOnce(i: number): Promise<string | null> {
    const src = this.src(i);
    const id = this.idOf(src);
    const items: { nonce: bigint; isBid: boolean; tick: number; qtyLots: bigint; window: WindowSpec }[] = [];
    const padKeys: ClientKey[] = [...this.tokenPadKeys()];
    for (let j = 0; j < ticksFor(i).length; j++) {
      const tick = ticksFor(i)[j];
      const nonce = NONCE_BASE + i * 1000 + j;
      items.push({
        nonce: BigInt(nonce),
        isBid: false,
        tick,
        qtyLots: BigInt(randInt(2, 5, this.rnd)),
        window: emptyRestWindow(),
      });
      padKeys.push({ t: "Level", market: MARKET, isBid: false, tick });
      padKeys.push({ t: "TickWord", market: MARKET, isBid: false, word: wordOf(tick) });
      padKeys.push({ t: "LevelPage", market: MARKET, isBid: false, tick, page: 0 });
    }
    padKeys.push({ t: "TickSummary", market: MARKET, isBid: false });
    padKeys.push({ t: "BestTick", market: MARKET, isBid: false });
    padKeys.push({ t: "BestTick", market: MARKET, isBid: true });
    for (const tick of ticksFor((i + 1) % this.a.accounts).slice(0, this.a.extraPad)) {
      padKeys.push({ t: "Level", market: MARKET, isBid: false, tick });
    }
    return this.submit(src, "replace_batch", () =>
      submitReplaceBatch(this.rpc, {
        contract: CONTRACT,
        secret: id.secret,
        owner: id.address,
        market: MARKET,
        items,
        padKeys,
        tokens: this.tokens,
      }),
    );
  }

  async run(): Promise<Record<string, unknown>> {
    const start = await this.latestFn();
    const end = start + this.a.ledgers;
    this.stop = false;
    const workers = Array.from({ length: this.a.accounts }, (_, i) => this.worker(i));
    await waitLedgers({
      latest: this.latestFn,
      start,
      count: this.a.ledgers,
      intervalMs: 10_000,
      progressEvery: 30,
      onProgress: (cur, e) =>
        this.record({
          acct: "stress",
          action: "progress",
          outcome: "tick",
          detail: `ledger ${cur}/${e} counts=${JSON.stringify(this.counts)}`,
        }),
      sleep: this.sleepFn,
      onPollError: () => undefined,
    });
    this.stop = true;
    await this.sleepFn(20_000);
    await Promise.all(workers);
    const summary = { start, end, counts: this.counts };
    this.record({ acct: "stress", action: "done", outcome: "summary", detail: JSON.stringify(summary) });
    process.stdout.write(JSON.stringify(summary, null, 1) + "\n");
    return summary;
  }

  private async worker(i: number): Promise<void> {
    while (!this.stop) {
      await this.batchOnce(i);
      const t0 = Date.now();
      while (Date.now() - t0 < this.a.pause * 1000 && !this.stop) {
        await this.sleepFn(Math.min(200, this.a.pause * 1000));
      }
    }
  }

  async analyze(): Promise<Record<string, unknown>> {
    const recs = [];
    for (const d of parseLogLines(this.readLog())) {
      if (d.action === "replace_batch" && d.outcome === "ok" && d.tx) recs.push(d);
    }
    process.stderr.write(`${recs.length} landed batches\n`);
    const got = await mapPool(recs, 8, async (r) => {
      try {
        const g = await this.rpc.getTransaction(String(r.tx));
        const fee = g.resultXdr ? decodeFeeCharged(g.resultXdr) : null;
        return { led: g.ledger ?? null, fee };
      } catch {
        return { led: null, fee: null };
      }
    });
    const per = new Map<number, { n: number; rw: number; wb: number; instr: number; fee: number }>();
    const delays: number[] = [];
    for (let i = 0; i < recs.length; i++) {
      const r = recs[i];
      const { led, fee } = got[i];
      if (led == null) continue;
      const p = per.get(led) ?? { n: 0, rw: 0, wb: 0, instr: 0, fee: 0 };
      p.n += 1;
      p.rw += Number(r.rw ?? 0);
      p.wb += Number(r.wb ?? 0);
      p.instr += Number(r.instr ?? 0);
      p.fee += fee ?? 0;
      per.set(led, p);
      if (r.submit_ledger != null) delays.push(led - Number(r.submit_ledger));
    }
    const vals = [...per.values()];
    const out = {
      ledgers_with_batches: per.size,
      per_ledger: {
        n: percentiles(vals.map((p) => p.n)),
        rw_declared: percentiles(vals.map((p) => p.rw)),
        write_bytes_declared: percentiles(vals.map((p) => p.wb)),
        instr_declared: percentiles(vals.map((p) => p.instr)),
      },
      inclusion_delay_ledgers: percentiles(delays),
      wall_s: percentiles(recs.map((r) => Number(r.wall_s ?? 0))),
      fee_charged: percentiles(got.map((g) => g.fee).filter((f): f is number => f != null)),
    };
    process.stdout.write(JSON.stringify(out, null, 1) + "\n");
    const top = [...per.entries()].sort((a, b) => b[1].wb - a[1].wb).slice(0, 10);
    process.stdout.write("busiest ledgers (by our declared write bytes):\n");
    for (const [led, p] of top) {
      process.stdout.write(
        `  ${led}: ${p.n} batches, rw ${p.rw} (${(p.rw / 10).toFixed(0)}% of cap), wb ${p.wb.toLocaleString("en-US")} (${(p.wb / 2867.2).toFixed(0)}% of cap), instr ${(p.instr / 1e6).toFixed(0)}M (${(p.instr / 5.8e6).toFixed(0)}% of cap)\n`,
      );
    }
    return out;
  }
}

export async function main(argv = process.argv.slice(2)): Promise<void> {
  const a = parseStressArgs(argv);
  if (a.network !== "testnet") throw new Error("only testnet is supported by the ops entry points today");
  const s = new Stress(a);
  if (a.phase === "seed") await s.seed();
  else if (a.phase === "run") await s.run();
  else await s.analyze();
}

const entry = process.argv[1] ? resolve(process.argv[1]) : "";
if (entry && fileURLToPath(import.meta.url) === entry) {
  main().catch((e) => {
    process.stderr.write(String(e) + "\n");
    process.exit(1);
  });
}